src/Microsoft.AspNetCore.Server.Kestrel/Internal/KestrelServerConfigureOptions.cs

```diff
@@ -0,0 +1,110 @@
+{
+    public class KestrelServerConfigureOptions : IConfigureOptions<KestrelServerOptions>
+    {
+        private const string DefaultCertificateSubjectName = "CN=localhost";
+        private const string DevelopmentSSLCertificateName = "localhost";
+
+        private readonly IHostingEnvironment _hostingEnvironment;
+        private readonly IConfiguration _configurationRoot;
+        private readonly ILoggerFactory _loggerFactory;
+
+        public KestrelServerConfigureOptions(
+            IHostingEnvironment hostingEnvironment,
+            IConfiguration configurationRoot,
+            ILoggerFactory loggerFactory)
+        {
+            _hostingEnvironment = hostingEnvironment;
+            _configurationRoot = configurationRoot;
+            _loggerFactory = loggerFactory;
+        }
+
+        public void Configure(KestrelServerOptions options)
+        {
+            BindConfiguration(options);
+        }
+
+        private void BindConfiguration(KestrelServerOptions options)
+        {
+            var certificateLoader = new CertificateLoader(_configurationRoot.GetSection("Certificates"), _loggerFactory, _hostingEnvironment.EnvironmentName);
+
+            foreach (var endPoint in _configurationRoot.GetSection("Microsoft:AspNetCore:Server:Kestrel:EndPoints").GetChildren())
+            {
+                BindEndPoint(options, endPoint, certificateLoader);
+            }
+        }
+
+        private void BindEndPoint(
+            KestrelServerOptions options,
+            IConfigurationSection endPoint,
+            CertificateLoader certificateLoader)
+        {
+            var configAddress = endPoint.GetValue<string>("Address");
+            var configPort = endPoint.GetValue<string>("Port");
+
+            if (!IPAddress.TryParse(configAddress, out var address))
+            {
+                throw new InvalidOperationException($"Invalid IP address in configuration: {configAddress}");
+            }
+
+            if (!int.TryParse(configPort, out var port))
+            {
+                throw new InvalidOperationException($"Invalid port in configuration: {configPort}");
+            }
+
+            options.Listen(address, port, listenOptions =>
+            {
+                var certificateConfig = endPoint.GetSection("Certificate");
+                X509Certificate2 certificate = null;
+                if (certificateConfig.Exists())
+                {
+                    try
+                    {
+                        try
+                        {
+                            certificate = certificateLoader.Load(certificateConfig).FirstOrDefault();
+                        }
+                        catch (KeyNotFoundException) when (certificateConfig.Value.Equals(DevelopmentSSLCertificateName, StringComparison.Ordinal) && _hostingEnvironment.IsDevelopment())
+                        {
+                            var storeLoader = new CertificateStoreLoader();
```

Comment:
I know you're just moving the code, but it still bothers me that we're newing this up here. I think loading the default cert should be a method on `CertificateLoader`. Thoughts?